# The goal is this assignment is to gain a basic introduction to Transformer model (BERT) and introduction to the Hugging Face library

## Install libraries

In [ ]:
!pip install git+https://github.com/huggingface/transformers.git
!pip install datasets
!pip install huggingface-hub

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Cloning https://github.com/huggingface/transformers.git to /tmp/pip-req-build-u3vl9n2m
  Running command git clone -q https://github.com/huggingface/transformers.git /tmp/pip-req-build-u3vl9n2m
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
     |████████████████████████████████| 163 kB 4.7 MB/s 
     |████████████████████████████████| 7.6 MB 42.6 MB/s 
  Created wheel for transformers: filename=transformers-4.25.0.dev0-py3-none-any.whl size=5645784 sha256=4910c663802783a61fb704872d53b4637359e90b8b6a69693efe04694c49b61e
  Stored in directory: /tmp/pip-ephem-wheel-cache-owrsrjaq/wheels/90/a5/44/6bcd83827c8a60628c5ad602f429cd5076bcce5f2a90054947
Successfully built transformers
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |███████████████████████████████

## Load SQUAD question answering dataset from the Hugging face library

In [ ]:
from datasets import load_dataset

datasets = load_dataset("squad")

Extracting data files:   0%|          | 0/2 [00:00<?, ?it/s]

Generating train split:   0%|          | 0/87599 [00:00<?, ? examples/s]

Generating validation split:   0%|          | 0/10570 [00:00<?, ? examples/s]

Dataset squad downloaded and prepared to /root/.cache/huggingface/datasets/squad/plain_text/1.0.0/d6ec3ceb99ca480ce37cdd35555d6cb2511d223b9150cce08a837ef62ffea453. Subsequent calls will reuse this data.


  0%|          | 0/2 [00:00<?, ?it/s]

In [ ]:
print(datasets["train"][0])

{'id': '5733be284776f41900661182', 'title': 'University_of_Notre_Dame', 'context': 'Architecturally, the school has a Catholic character. Atop the Main Building\'s gold dome is a golden statue of the Virgin Mary. Immediately in front of the Main Building and facing it, is a copper statue of Christ with arms upraised with the legend "Venite Ad Me Omnes". Next to the Main Building is the Basilica of the Sacred Heart. Immediately behind the basilica is the Grotto, a Marian place of prayer and reflection. It is a replica of the grotto at Lourdes, France where the Virgin Mary reputedly appeared to Saint Bernadette Soubirous in 1858. At the end of the main drive (and in a direct line that connects through 3 statues and the Gold Dome), is a simple, modern stone statue of Mary.', 'question': 'To whom did the Virgin Mary allegedly appear in 1858 in Lourdes France?', 'answers': {'text': ['Saint Bernadette Soubirous'], 'answer_start': [515]}}


In [ ]:
print(datasets["train"][0]['context'])

Architecturally, the school has a Catholic character. Atop the Main Building's gold dome is a golden statue of the Virgin Mary. Immediately in front of the Main Building and facing it, is a copper statue of Christ with arms upraised with the legend "Venite Ad Me Omnes". Next to the Main Building is the Basilica of the Sacred Heart. Immediately behind the basilica is the Grotto, a Marian place of prayer and reflection. It is a replica of the grotto at Lourdes, France where the Virgin Mary reputedly appeared to Saint Bernadette Soubirous in 1858. At the end of the main drive (and in a direct line that connects through 3 statues and the Gold Dome), is a simple, modern stone statue of Mary.


In [ ]:
print(datasets["train"][0]['question'])

To whom did the Virgin Mary allegedly appear in 1858 in Lourdes France?


In [ ]:
print(datasets["train"][0]['answers'])

{'text': ['Saint Bernadette Soubirous'], 'answer_start': [515]}


## Problem 1: Use “distilbert-base-cased-distilled-squad” for question-answering on a context and questions that you specify. Use the model to answer 10 differentquestions. 

In [ ]:
#import random
#Generate 10 random numbers between 0 and 30
#randomlist = random.sample(range(0, 30), 10)
#print(randomlist)

In [ ]:
# Choose 10 random contexts from the SQUAD training dataset
randomlist = [14, 4, 16, 18, 0, 23, 22, 24, 27, 7]

#### Create an instance of the pretrained model

In [ ]:
from transformers import pipeline
question_answerer = pipeline("question-answering", model='distilbert-base-cased-distilled-squad')

Downloading:   0%|          | 0.00/473 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/261M [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/29.0 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/213k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/436k [00:00<?, ?B/s]

#### Define context. Formulate a question. Then let the model predict the answer to the question. 

In [ ]:
context1 = (datasets["train"][randomlist[0]]['context'])
print(context1)

The university is the major seat of the Congregation of Holy Cross (albeit not its official headquarters, which are in Rome). Its main seminary, Moreau Seminary, is located on the campus across St. Joseph lake from the Main Building. Old College, the oldest building on campus and located near the shore of St. Mary lake, houses undergraduate seminarians. Retired priests and brothers reside in Fatima House (a former retreat center), Holy Cross House, as well as Columba Hall near the Grotto. The university through the Moreau Seminary has ties to theologian Frederick Buechner. While not Catholic, Buechner has praised writers from Notre Dame and Moreau Seminary created a Buechner Prize for Preaching.


In [ ]:
question1 = "Where is main seminary located?"
result1 = question_answerer(question=question1,context=context1)
print(f"Answer: '{result1['answer']}', score: {round(result1['score'], 4)}, start: {result1['start']}, end: {result1['end']}")

Answer: 'on the campus across St. Joseph lake', score: 0.2403, start: 173, end: 209


In [ ]:
context2 = (datasets["train"][randomlist[1]]['context'])
print(context2)

Architecturally, the school has a Catholic character. Atop the Main Building's gold dome is a golden statue of the Virgin Mary. Immediately in front of the Main Building and facing it, is a copper statue of Christ with arms upraised with the legend "Venite Ad Me Omnes". Next to the Main Building is the Basilica of the Sacred Heart. Immediately behind the basilica is the Grotto, a Marian place of prayer and reflection. It is a replica of the grotto at Lourdes, France where the Virgin Mary reputedly appeared to Saint Bernadette Soubirous in 1858. At the end of the main drive (and in a direct line that connects through 3 statues and the Gold Dome), is a simple, modern stone statue of Mary.


In [ ]:
question2 = "What type of architecture does the school have?"
result2 = question_answerer(question=question2,     context=context2)
print(f"Answer: '{result2['answer']}', score: {round(result2['score'], 4)}, start: {result2['start']}, end: {result2['end']}")

Answer: 'Catholic', score: 0.8411, start: 34, end: 42


In [ ]:
context3 = (datasets["train"][randomlist[2]]['context'])
print(context3)

The College of Engineering was established in 1920, however, early courses in civil and mechanical engineering were a part of the College of Science since the 1870s. Today the college, housed in the Fitzpatrick, Cushing, and Stinson-Remick Halls of Engineering, includes five departments of study – aerospace and mechanical engineering, chemical and biomolecular engineering, civil engineering and geological sciences, computer science and engineering, and electrical engineering – with eight B.S. degrees offered. Additionally, the college offers five-year dual degree programs with the Colleges of Arts and Letters and of Business awarding additional B.A. and Master of Business Administration (MBA) degrees, respectively.


In [ ]:
question3 = "What departments of study are in the Collge of Engineering today?"
result3 = question_answerer(question=question3,     context=context3)
print(f"Answer: '{result3['answer']}', score: {round(result3['score'], 4)}, start: {result3['start']}, end: {result3['end']}")

Answer: 'aerospace and mechanical engineering, chemical and biomolecular engineering', score: 0.0408, start: 299, end: 374


In [ ]:
context4 = (datasets["train"][randomlist[3]]['context'])
print(context4)

The College of Engineering was established in 1920, however, early courses in civil and mechanical engineering were a part of the College of Science since the 1870s. Today the college, housed in the Fitzpatrick, Cushing, and Stinson-Remick Halls of Engineering, includes five departments of study – aerospace and mechanical engineering, chemical and biomolecular engineering, civil engineering and geological sciences, computer science and engineering, and electrical engineering – with eight B.S. degrees offered. Additionally, the college offers five-year dual degree programs with the Colleges of Arts and Letters and of Business awarding additional B.A. and Master of Business Administration (MBA) degrees, respectively.


In [ ]:
question4 = "Where is the College of Engineering located?"
result4 = question_answerer(question=question4,     context=context4)
print(f"Answer: '{result4['answer']}', score: {round(result4['score'], 4)}, start: {result4['start']}, end: {result4['end']}")

Answer: 'Stinson-Remick Halls of Engineering', score: 0.0583, start: 225, end: 260


In [ ]:
context5 = (datasets["train"][randomlist[4]]['context'])
print(context5)

Architecturally, the school has a Catholic character. Atop the Main Building's gold dome is a golden statue of the Virgin Mary. Immediately in front of the Main Building and facing it, is a copper statue of Christ with arms upraised with the legend "Venite Ad Me Omnes". Next to the Main Building is the Basilica of the Sacred Heart. Immediately behind the basilica is the Grotto, a Marian place of prayer and reflection. It is a replica of the grotto at Lourdes, France where the Virgin Mary reputedly appeared to Saint Bernadette Soubirous in 1858. At the end of the main drive (and in a direct line that connects through 3 statues and the Gold Dome), is a simple, modern stone statue of Mary.


In [ ]:
question5 = "Where is the golden statue of Virgin Mary placed?"
result5 = question_answerer(question=question5,     context=context5)
print(f"Answer: '{result5['answer']}', score: {round(result5['score'], 4)}, start: {result5['start']}, end: {result5['end']}")

Answer: 'Atop the Main Building', score: 0.3605, start: 54, end: 76


In [ ]:
context6 = (datasets["train"][randomlist[5]]['context'])
print(context6)

All of Notre Dame's undergraduate students are a part of one of the five undergraduate colleges at the school or are in the First Year of Studies program. The First Year of Studies program was established in 1962 to guide incoming freshmen in their first year at the school before they have declared a major. Each student is given an academic advisor from the program who helps them to choose classes that give them exposure to any major in which they are interested. The program also includes a Learning Resource Center which provides time management, collaborative learning, and subject tutoring. This program has been recognized previously, by U.S. News & World Report, as outstanding.


In [ ]:
question6 = "When was the first year of studies program established at Notre Dame?"
result6 = question_answerer(question=question6,     context=context6)
print(f"Answer: '{result6['answer']}', score: {round(result6['score'], 4)}, start: {result6['start']}, end: {result6['end']}")

Answer: '1962', score: 0.9926, start: 208, end: 212


In [ ]:
context7 = (datasets["train"][randomlist[6]]['context'])
print(context7)

All of Notre Dame's undergraduate students are a part of one of the five undergraduate colleges at the school or are in the First Year of Studies program. The First Year of Studies program was established in 1962 to guide incoming freshmen in their first year at the school before they have declared a major. Each student is given an academic advisor from the program who helps them to choose classes that give them exposure to any major in which they are interested. The program also includes a Learning Resource Center which provides time management, collaborative learning, and subject tutoring. This program has been recognized previously, by U.S. News & World Report, as outstanding.


In [ ]:
question7 = "What is the role of the academic advisor?"
result7 = question_answerer(question=question7,     context=context7)
print(f"Answer: '{result7['answer']}', score: {round(result7['score'], 4)}, start: {result7['start']}, end: {result7['end']}")

Answer: 'helps them to choose classes that give them exposure to any major', score: 0.0798, start: 372, end: 437


In [ ]:
context8 = (datasets["train"][randomlist[7]]['context'])
print(context8)

The university first offered graduate degrees, in the form of a Master of Arts (MA), in the 1854–1855 academic year. The program expanded to include Master of Laws (LL.M.) and Master of Civil Engineering in its early stages of growth, before a formal graduate school education was developed with a thesis not required to receive the degrees. This changed in 1924 with formal requirements developed for graduate degrees, including offering Doctorate (PhD) degrees. Today each of the five colleges offer graduate education. Most of the departments from the College of Arts and Letters offer PhD programs, while a professional Master of Divinity (M.Div.) program also exists. All of the departments in the College of Science offer PhD programs, except for the Department of Pre-Professional Studies. The School of Architecture offers a Master of Architecture, while each of the departments of the College of Engineering offer PhD programs. The College of Business offers multiple professional programs i

In [ ]:
question8 = "What changed in 1924?"
result8 = question_answerer(question=question8,     context=context8)
print(f"Answer: '{result8['answer']}', score: {round(result8['score'], 4)}, start: {result8['start']}, end: {result8['end']}")

Answer: 'formal requirements developed for graduate degrees', score: 0.5206, start: 368, end: 418


In [ ]:
context9 = (datasets["train"][randomlist[8]]['context'])
print(context9)

The university first offered graduate degrees, in the form of a Master of Arts (MA), in the 1854–1855 academic year. The program expanded to include Master of Laws (LL.M.) and Master of Civil Engineering in its early stages of growth, before a formal graduate school education was developed with a thesis not required to receive the degrees. This changed in 1924 with formal requirements developed for graduate degrees, including offering Doctorate (PhD) degrees. Today each of the five colleges offer graduate education. Most of the departments from the College of Arts and Letters offer PhD programs, while a professional Master of Divinity (M.Div.) program also exists. All of the departments in the College of Science offer PhD programs, except for the Department of Pre-Professional Studies. The School of Architecture offers a Master of Architecture, while each of the departments of the College of Engineering offer PhD programs. The College of Business offers multiple professional programs i

In [ ]:
question9 = "When was Master of Civil Engineering added?"
result9 = question_answerer(question=question9,     context=context9)
print(f"Answer: '{result9['answer']}', score: {round(result9['score'], 4)}, start: {result9['start']}, end: {result9['end']}")

Answer: 'early stages of growth', score: 0.6819, start: 211, end: 233


In [ ]:
context10 = (datasets["train"][randomlist[9]]['context'])
print(context10)

As at most other universities, Notre Dame's students run a number of news media outlets. The nine student-run outlets include three newspapers, both a radio and television station, and several magazines and journals. Begun as a one-page journal in September 1876, the Scholastic magazine is issued twice monthly and claims to be the oldest continuous collegiate publication in the United States. The other magazine, The Juggler, is released twice a year and focuses on student literature and artwork. The Dome yearbook is published annually. The newspapers have varying publication interests, with The Observer published daily and mainly reporting university and other news, and staffed by students from both Notre Dame and Saint Mary's College. Unlike Scholastic and The Dome, The Observer is an independent publication and does not have a faculty advisor or any editorial oversight from the University. In 1987, when some students believed that The Observer began to show a conservative bias, a lib

In [ ]:
question10 = "How often is Scholastic magazine issued?"
result10 = question_answerer(question=question10,     context=context10)
print(f"Answer: '{result10['answer']}', score: {round(result10['score'], 4)}, start: {result10['start']}, end: {result10['end']}")

Answer: 'twice monthly', score: 0.9307, start: 298, end: 311


## Problem 2: Use the “distilbert-base-cased” transformer for question answering. To do so, you need to fine tune the model on a question-answering dataset (SQuAD dataset). 

### Data Preprocessing

In [ ]:
from transformers import AutoTokenizer
model_checkpoint = "distilbert-base-cased"
tokenizer = AutoTokenizer.from_pretrained(model_checkpoint)

Downloading:   0%|          | 0.00/29.0 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/411 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/213k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/436k [00:00<?, ?B/s]

In [ ]:
max_length = 384  # The maximum length of a feature (question and context)
doc_stride = (
    128  # The authorized overlap between two part of the context when splitting
)

In [ ]:
def prepare_train_features(examples):
    # Tokenize our examples with truncation and padding, but keep the overflows using a
    # stride. This results in one example possible giving several features when a context is long,
    # each of those features having a context that overlaps a bit the context of the previous
    # feature.
    examples["question"] = [q.lstrip() for q in examples["question"]]
    examples["context"] = [c.lstrip() for c in examples["context"]]
    tokenized_examples = tokenizer(
        examples["question"],
        examples["context"],
        truncation="only_second",
        max_length=max_length,
        stride=doc_stride,
        return_overflowing_tokens=True,
        return_offsets_mapping=True,
        padding="max_length",
    )

    # Since one example might give us several features if it has a long context, we need a
    # map from a feature to its corresponding example. This key gives us just that.
    sample_mapping = tokenized_examples.pop("overflow_to_sample_mapping")
    # The offset mappings will give us a map from token to character position in the original
    # context. This will help us compute the start_positions and end_positions.
    offset_mapping = tokenized_examples.pop("offset_mapping")

    # Let's label those examples!
    tokenized_examples["start_positions"] = []
    tokenized_examples["end_positions"] = []

    for i, offsets in enumerate(offset_mapping):
        # We will label impossible answers with the index of the CLS token.
        input_ids = tokenized_examples["input_ids"][i]
        cls_index = input_ids.index(tokenizer.cls_token_id)

        # Grab the sequence corresponding to that example (to know what is the context and what
        # is the question).
        sequence_ids = tokenized_examples.sequence_ids(i)

        # One example can give several spans, this is the index of the example containing this
        # span of text.
        sample_index = sample_mapping[i]
        answers = examples["answers"][sample_index]
        # If no answers are given, set the cls_index as answer.
        if len(answers["answer_start"]) == 0:
            tokenized_examples["start_positions"].append(cls_index)
            tokenized_examples["end_positions"].append(cls_index)
        else:
            # Start/end character index of the answer in the text.
            start_char = answers["answer_start"][0]
            end_char = start_char + len(answers["text"][0])

            # Start token index of the current span in the text.
            token_start_index = 0
            while sequence_ids[token_start_index] != 1:
                token_start_index += 1

            # End token index of the current span in the text.
            token_end_index = len(input_ids) - 1
            while sequence_ids[token_end_index] != 1:
                token_end_index -= 1

            # Detect if the answer is out of the span (in which case this feature is labeled with the
            # CLS index).
            if not (
                offsets[token_start_index][0] <= start_char
                and offsets[token_end_index][1] >= end_char
            ):
                tokenized_examples["start_positions"].append(cls_index)
                tokenized_examples["end_positions"].append(cls_index)
            else:
                # Otherwise move the token_start_index and token_end_index to the two ends of the
                # answer.
                # Note: we could go after the last offset if the answer is the last word (edge
                # case).
                while (
                    token_start_index < len(offsets)
                    and offsets[token_start_index][0] <= start_char
                ):
                    token_start_index += 1
                tokenized_examples["start_positions"].append(token_start_index - 1)
                while offsets[token_end_index][1] >= end_char:
                    token_end_index -= 1
                tokenized_examples["end_positions"].append(token_end_index + 1)

    return tokenized_examples

In [ ]:
tokenized_datasets = datasets.map(
    prepare_train_features,
    batched=True,
    remove_columns=datasets["train"].column_names,
    num_proc=3,
)

#0:   0%|          | 0/30 [00:00<?, ?ba/s]

#1:   0%|          | 0/30 [00:00<?, ?ba/s]

#2:   0%|          | 0/30 [00:00<?, ?ba/s]

#0:   0%|          | 0/4 [00:00<?, ?ba/s]

#2:   0%|          | 0/4 [00:00<?, ?ba/s]

#1:   0%|          | 0/4 [00:00<?, ?ba/s]

In [ ]:
train_set = tokenized_datasets["train"].with_format("numpy")[
    :
]  # Load the whole dataset as a dict of numpy arrays
validation_set = tokenized_datasets["validation"].with_format("numpy")[:]

### Fine-tuning the model

In [ ]:
from transformers import TFAutoModelForQuestionAnswering

model = TFAutoModelForQuestionAnswering.from_pretrained(model_checkpoint)

Downloading:   0%|          | 0.00/354M [00:00<?, ?B/s]

Some layers from the model checkpoint at distilbert-base-cased were not used when initializing TFDistilBertForQuestionAnswering: ['vocab_layer_norm', 'vocab_transform', 'activation_13', 'vocab_projector']
- This IS expected if you are initializing TFDistilBertForQuestionAnswering from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing TFDistilBertForQuestionAnswering from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some layers of TFDistilBertForQuestionAnswering were not initialized from the model checkpoint at distilbert-base-cased and are newly initialized: ['qa_outputs', 'dropout_19']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


In [ ]:
import tensorflow as tf
from tensorflow import keras

optimizer = keras.optimizers.Adam(learning_rate=5e-5)

In [ ]:
# Optionally uncomment the next line for float16 training
keras.mixed_precision.set_global_policy("mixed_float16")

model.compile(optimizer=optimizer)

No loss specified in compile() - the model's internal loss computation will be used as the loss. Don't panic - this is a common way to train TensorFlow models in Transformers! To disable this behaviour please pass a loss argument, or explicitly pass `loss=None` if you do not want your model to compute a loss.


In [ ]:
# Train the model
model.fit(train_set, validation_data=validation_set, epochs=1, batch_size=20)

4437/4437 [==============================] - 482s 109ms/step - loss: 0.3144 - val_loss: 1.6736


### Test the model

In [ ]:
# Test the model on the same questions as in Problem 1
inputs = tokenizer([context1], [question1], return_tensors="np")
outputs = model(inputs)
start_position = tf.argmax(outputs.start_logits, axis=1)
end_position = tf.argmax(outputs.end_logits, axis=1)
print(int(start_position), int(end_position[0]))

answer = inputs["input_ids"][0, int(start_position) : int(end_position) + 1]
print(answer)

0 0
[101]


In [ ]:
print(context1)

The university is the major seat of the Congregation of Holy Cross (albeit not its official headquarters, which are in Rome). Its main seminary, Moreau Seminary, is located on the campus across St. Joseph lake from the Main Building. Old College, the oldest building on campus and located near the shore of St. Mary lake, houses undergraduate seminarians. Retired priests and brothers reside in Fatima House (a former retreat center), Holy Cross House, as well as Columba Hall near the Grotto. The university through the Moreau Seminary has ties to theologian Frederick Buechner. While not Catholic, Buechner has praised writers from Notre Dame and Moreau Seminary created a Buechner Prize for Preaching.


In [ ]:
print(question1)

Where is main seminary located?


In [ ]:
print(tokenizer.decode(answer))

[CLS]


In [ ]:
inputs = tokenizer([context2], [question2], return_tensors="np")
outputs = model(inputs)
start_position = tf.argmax(outputs.start_logits, axis=1)
end_position = tf.argmax(outputs.end_logits, axis=1)
print(int(start_position), int(end_position[0]))

answer = inputs["input_ids"][0, int(start_position) : int(end_position) + 1]
print(answer)

0 0
[101]


In [ ]:
print(context2)
print(question2)

Architecturally, the school has a Catholic character. Atop the Main Building's gold dome is a golden statue of the Virgin Mary. Immediately in front of the Main Building and facing it, is a copper statue of Christ with arms upraised with the legend "Venite Ad Me Omnes". Next to the Main Building is the Basilica of the Sacred Heart. Immediately behind the basilica is the Grotto, a Marian place of prayer and reflection. It is a replica of the grotto at Lourdes, France where the Virgin Mary reputedly appeared to Saint Bernadette Soubirous in 1858. At the end of the main drive (and in a direct line that connects through 3 statues and the Gold Dome), is a simple, modern stone statue of Mary.
What type of architecture does the school have?


In [ ]:
print(tokenizer.decode(answer))

[CLS]


In [ ]:
inputs = tokenizer([context3], [question3], return_tensors="np")
outputs = model(inputs)
start_position = tf.argmax(outputs.start_logits, axis=1)
end_position = tf.argmax(outputs.end_logits, axis=1)
print(int(start_position), int(end_position[0]))

answer = inputs["input_ids"][0, int(start_position) : int(end_position) + 1]
print(answer)

24 90
[ 1531  1104  2444  1290  1103 15583   119  3570  1103  2134   117  6960
  1107  1103 17355  5745 17015   117   140 17760   117  1105  1457  4935
  1320   118 11336  7257  1377  1944  1116  1104  3939   117  2075  1421
  7844  1104  2025   782 27689  1105  6676  3752   117  5297  1105 25128
  3702  1513 11702  3752   117  2987  3752  1105 16622  8614   117  2775
  2598  1105  3752   117  1105  6538  3752]


In [ ]:
print(context3)
print(question3)

The College of Engineering was established in 1920, however, early courses in civil and mechanical engineering were a part of the College of Science since the 1870s. Today the college, housed in the Fitzpatrick, Cushing, and Stinson-Remick Halls of Engineering, includes five departments of study – aerospace and mechanical engineering, chemical and biomolecular engineering, civil engineering and geological sciences, computer science and engineering, and electrical engineering – with eight B.S. degrees offered. Additionally, the college offers five-year dual degree programs with the Colleges of Arts and Letters and of Business awarding additional B.A. and Master of Business Administration (MBA) degrees, respectively.
What departments of study are in the Collge of Engineering today?


In [ ]:
print(tokenizer.decode(answer))

College of Science since the 1870s. Today the college, housed in the Fitzpatrick, Cushing, and Stinson - Remick Halls of Engineering, includes five departments of study – aerospace and mechanical engineering, chemical and biomolecular engineering, civil engineering and geological sciences, computer science and engineering, and electrical engineering


In [ ]:
inputs = tokenizer([context4], [question4], return_tensors="np")
outputs = model(inputs)
start_position = tf.argmax(outputs.start_logits, axis=1)
end_position = tf.argmax(outputs.end_logits, axis=1)
print(int(start_position), int(end_position[0]))

answer = inputs["input_ids"][0, int(start_position) : int(end_position) + 1]
print(answer)

24 56
[ 1531  1104  2444  1290  1103 15583   119  3570  1103  2134   117  6960
  1107  1103 17355  5745 17015   117   140 17760   117  1105  1457  4935
  1320   118 11336  7257  1377  1944  1116  1104  3939]


In [ ]:
print(context4)
print(question4)

The College of Engineering was established in 1920, however, early courses in civil and mechanical engineering were a part of the College of Science since the 1870s. Today the college, housed in the Fitzpatrick, Cushing, and Stinson-Remick Halls of Engineering, includes five departments of study – aerospace and mechanical engineering, chemical and biomolecular engineering, civil engineering and geological sciences, computer science and engineering, and electrical engineering – with eight B.S. degrees offered. Additionally, the college offers five-year dual degree programs with the Colleges of Arts and Letters and of Business awarding additional B.A. and Master of Business Administration (MBA) degrees, respectively.
Where is the College of Engineering located?


In [ ]:
print(tokenizer.decode(answer))

College of Science since the 1870s. Today the college, housed in the Fitzpatrick, Cushing, and Stinson - Remick Halls of Engineering


In [ ]:
inputs = tokenizer([context5], [question5], return_tensors="np")
outputs = model(inputs)
start_position = tf.argmax(outputs.start_logits, axis=1)
end_position = tf.argmax(outputs.end_logits, axis=1)
print(int(start_position), int(end_position[0]))

answer = inputs["input_ids"][0, int(start_position) : int(end_position) + 1]
print(answer)

0 0
[101]


In [ ]:
print(context5)
print(question5)

Architecturally, the school has a Catholic character. Atop the Main Building's gold dome is a golden statue of the Virgin Mary. Immediately in front of the Main Building and facing it, is a copper statue of Christ with arms upraised with the legend "Venite Ad Me Omnes". Next to the Main Building is the Basilica of the Sacred Heart. Immediately behind the basilica is the Grotto, a Marian place of prayer and reflection. It is a replica of the grotto at Lourdes, France where the Virgin Mary reputedly appeared to Saint Bernadette Soubirous in 1858. At the end of the main drive (and in a direct line that connects through 3 statues and the Gold Dome), is a simple, modern stone statue of Mary.
Where is the golden statue of Virgin Mary placed?


In [ ]:
print(tokenizer.decode(answer))

[CLS]


In [ ]:
inputs = tokenizer([context6], [question6], return_tensors="np")
outputs = model(inputs)
start_position = tf.argmax(outputs.start_logits, axis=1)
end_position = tf.argmax(outputs.end_logits, axis=1)
print(int(start_position), int(end_position[0]))

answer = inputs["input_ids"][0, int(start_position) : int(end_position) + 1]
print(answer)

26 0
[]


In [ ]:
print(context6)
print(question6)

All of Notre Dame's undergraduate students are a part of one of the five undergraduate colleges at the school or are in the First Year of Studies program. The First Year of Studies program was established in 1962 to guide incoming freshmen in their first year at the school before they have declared a major. Each student is given an academic advisor from the program who helps them to choose classes that give them exposure to any major in which they are interested. The program also includes a Learning Resource Center which provides time management, collaborative learning, and subject tutoring. This program has been recognized previously, by U.S. News & World Report, as outstanding.
When was the first year of studies program established at Notre Dame?


In [ ]:
print(tokenizer.decode(answer))

In [ ]:
inputs = tokenizer([context7], [question7], return_tensors="np")
outputs = model(inputs)
start_position = tf.argmax(outputs.start_logits, axis=1)
end_position = tf.argmax(outputs.end_logits, axis=1)
print(int(start_position), int(end_position[0]))

answer = inputs["input_ids"][0, int(start_position) : int(end_position) + 1]
print(answer)

0 0
[101]


In [ ]:
print(context7)
print(question7)

All of Notre Dame's undergraduate students are a part of one of the five undergraduate colleges at the school or are in the First Year of Studies program. The First Year of Studies program was established in 1962 to guide incoming freshmen in their first year at the school before they have declared a major. Each student is given an academic advisor from the program who helps them to choose classes that give them exposure to any major in which they are interested. The program also includes a Learning Resource Center which provides time management, collaborative learning, and subject tutoring. This program has been recognized previously, by U.S. News & World Report, as outstanding.
What is the role of the academic advisor?


In [ ]:
print(tokenizer.decode(answer))

[CLS]


In [ ]:
inputs = tokenizer([context8], [question8], return_tensors="np")
outputs = model(inputs)
start_position = tf.argmax(outputs.start_logits, axis=1)
end_position = tf.argmax(outputs.end_logits, axis=1)
print(int(start_position), int(end_position[0]))

answer = inputs["input_ids"][0, int(start_position) : int(end_position) + 1]
print(answer)

13 90
[ 3257  1104  2334   113  9960   114   117  1107  1103  8023   782  8082
  3397  1214   119  1109  1788  3631  1106  1511  3257  1104 12077   113
 12427   119   150   119   114  1105  3257  1104  3145  3939  1107  1157
  1346  5251  1104  3213   117  1196   170  4698  4469  1278  1972  1108
  1872  1114   170  9593  1136  2320  1106  3531  1103  4842   119  1188
  2014  1107  4002  1114  4698  5420  1872  1111  4469  4842   117  1259
  4733 24893   113  7735   114  4842]


In [ ]:
print(context8)
print(question8)

The university first offered graduate degrees, in the form of a Master of Arts (MA), in the 1854–1855 academic year. The program expanded to include Master of Laws (LL.M.) and Master of Civil Engineering in its early stages of growth, before a formal graduate school education was developed with a thesis not required to receive the degrees. This changed in 1924 with formal requirements developed for graduate degrees, including offering Doctorate (PhD) degrees. Today each of the five colleges offer graduate education. Most of the departments from the College of Arts and Letters offer PhD programs, while a professional Master of Divinity (M.Div.) program also exists. All of the departments in the College of Science offer PhD programs, except for the Department of Pre-Professional Studies. The School of Architecture offers a Master of Architecture, while each of the departments of the College of Engineering offer PhD programs. The College of Business offers multiple professional programs i

In [ ]:
print(tokenizer.decode(answer))

Master of Arts ( MA ), in the 1854 – 1855 academic year. The program expanded to include Master of Laws ( LL. M. ) and Master of Civil Engineering in its early stages of growth, before a formal graduate school education was developed with a thesis not required to receive the degrees. This changed in 1924 with formal requirements developed for graduate degrees, including offering Doctorate ( PhD ) degrees


In [ ]:
inputs = tokenizer([context9], [question9], return_tensors="np")
outputs = model(inputs)
start_position = tf.argmax(outputs.start_logits, axis=1)
end_position = tf.argmax(outputs.end_logits, axis=1)
print(int(start_position), int(end_position[0]))

answer = inputs["input_ids"][0, int(start_position) : int(end_position) + 1]
print(answer)

75 75
[4002]


In [ ]:
print(context9)
print(question9)

The university first offered graduate degrees, in the form of a Master of Arts (MA), in the 1854–1855 academic year. The program expanded to include Master of Laws (LL.M.) and Master of Civil Engineering in its early stages of growth, before a formal graduate school education was developed with a thesis not required to receive the degrees. This changed in 1924 with formal requirements developed for graduate degrees, including offering Doctorate (PhD) degrees. Today each of the five colleges offer graduate education. Most of the departments from the College of Arts and Letters offer PhD programs, while a professional Master of Divinity (M.Div.) program also exists. All of the departments in the College of Science offer PhD programs, except for the Department of Pre-Professional Studies. The School of Architecture offers a Master of Architecture, while each of the departments of the College of Engineering offer PhD programs. The College of Business offers multiple professional programs i

In [ ]:
print(tokenizer.decode(answer))

1924


In [ ]:
inputs = tokenizer([context10], [question10], return_tensors="np")
outputs = model(inputs)
start_position = tf.argmax(outputs.start_logits, axis=1)
end_position = tf.argmax(outputs.end_logits, axis=1)
print(int(start_position), int(end_position[0]))

answer = inputs["input_ids"][0, int(start_position) : int(end_position) + 1]
print(answer)

0 0
[101]


In [ ]:
print(context10)
print(question10)

As at most other universities, Notre Dame's students run a number of news media outlets. The nine student-run outlets include three newspapers, both a radio and television station, and several magazines and journals. Begun as a one-page journal in September 1876, the Scholastic magazine is issued twice monthly and claims to be the oldest continuous collegiate publication in the United States. The other magazine, The Juggler, is released twice a year and focuses on student literature and artwork. The Dome yearbook is published annually. The newspapers have varying publication interests, with The Observer published daily and mainly reporting university and other news, and staffed by students from both Notre Dame and Saint Mary's College. Unlike Scholastic and The Dome, The Observer is an independent publication and does not have a faculty advisor or any editorial oversight from the University. In 1987, when some students believed that The Observer began to show a conservative bias, a lib

In [ ]:
print(tokenizer.decode(answer))

[CLS]


## Problem 3: Compare the results in (step 1) and (step 2) with elaboration.

To summarize, the answers to questions in Problem 1 are below:
1. 'on the campus across St. Joseph lake', score: 0.2403
2. 'Catholic', score: 0.8411
3. 'aerospace and mechanical engineering, chemical and biomolecular engineering', score: 0.0408
4. 'Stinson-Remick Halls of Engineering', score: 0.0583 
5. 'Atop the Main Building', score: 0.3605
6. '1962', score: 0.9926
7. 'helps them to choose classes that give them exposure to any major', score: 0.0798 
8. 'formal requirements developed for graduate degrees', score: 0.5206 
9. 'early stages of growth', score: 0.6819  
10. 'twice monthly', score: 0.9307  


The answers to the same questions in Problem 2 are below:

1. [CLS]
2. [CLS]
3. College of Science since the 1870s. Today the college, housed in the Fitzpatrick, Cushing, and Stinson - Remick Halls of Engineering, includes five departments of study – aerospace and mechanical engineering, chemical and biomolecular engineering, civil engineering and geological sciences, computer science and engineering, and electrical engineering
4. College of Science since the 1870s. Today the college, housed in the Fitzpatrick, Cushing, and Stinson - Remick Halls of Engineering
5. [CLS]
6. Blank
7. [CLS]
8. Master of Arts ( MA ), in the 1854 – 1855 academic year. The program expanded to include Master of Laws ( LL. M. ) and Master of Civil Engineering in its early stages of growth, before a formal graduate school education was developed with a thesis not required to receive the degrees. This changed in 1924 with formal requirements developed for graduate degrees, including offering Doctorate ( PhD ) degrees
9. 1924
10. [CLS]

ANALYSIS:
The expectation was that both steps would yield similar results since in Problem 1, **distilbert-base-cased-distilled-squad** was used, which is a fine tuned version of **DistilBERT base model (cased)** model trained on the SQUAD question answering dataset. The second step uses the **DistilBERT base model (cased)** model and then fine tunes it with the same SQUAD dataset. 

However, the results from the two steps are very diifferent. The **distilbert-base-cased-distilled-squad** model gave correct answers for all the questions. Whereas the **DistilBERT base model (cased)** finetuned on SQUAD dataset didn't perform very well. In most cases it returned [CLS] token itself. The results didn't change with increasing the number of training epochs. 

One difference could be that the pre-processing steps are different between the  **distilbert-base-cased-distilled-squad** and fine tuned version of **DistilBERT base model (cased)**.  Secondly, the training differences (hyperparameters used during fine tuning) may have lead to different results also.

# References

1. Hugging Face Model Card for **DistilBERT base model (cased)** . DistilBERT English language model pretrained on the same data used to pretrain Bert (concatenation of the Toronto Book Corpus and full English Wikipedia) using distillation with the supervision of the bert-base-cased version of Bert. The model has 6 layers, 768 dimension and 12 heads, totalizing 65M parameters.  *It is intended to be fine-tuned on a downstream task*. https://huggingface.co/distilbert-base-cased
2. Hugging Face Model details for **distilbert-base-cased-distilled-squad**. A finetuned version of distilbert-base-cased finetuned using (a second step of) knowledge distillation on SQuAD 1.0. This model reaches a F1 score of 87.1 on the dev set (for comparison, Bert bert-base-cased version reaches a 88.7 F1 score). https://huggingface.co/distilbert-base-cased-distilled-squad
3. Explanation of how to train a DistillBER. https://github.com/huggingface/transformers/tree/main/examples/research_projects/distillation
4. Squad dataset. Stanford Question Answering Dataset (SQuAD) is a reading comprehension dataset, consisting of questions posed by crowdworkers on a set of Wikipedia articles, where the answer to every question is a segment of text, or span, from the corresponding reading passage, or the question might be unanswerable. https://huggingface.co/datasets/squad
5. Keras question-answering example. https://keras.io/examples/nlp/question_answering/
  

*   Context: Input document
*   Question: a question about the document
*   Answer: an answer to the question from a given context



